### ISSUE TITLE
The cli should respect XDG spec for config and credential file storage

### ISSUE THREAD

**Issue by jasonkarns** [aws/aws-cli#2433]
The XDG spec defines where config files and credential files should be placed.

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Instead of writing to $HOME, the aws config directory should be `$XDG_CONFIG_HOME/aws`.

**Comment by stealthycoin** [comment]
I'll mark this as a feature request. It will require some discussion.

**Comment by qrevel** [comment]
It should also respect the XDG cache folder when assuming role.

**Comment by ASayre** [comment]
Good Morning!

We're closing this issue here on GitHub, as part of our migration to [UserVoice](https://aws.uservoice.com/forums/598381-aws-command-line-interface) for feature requests involving the AWS CLI.

This will let us get the most important features to you, by making it easier to search for and show support for the features you care the most about, without diluting the conversation with bug reports.

As a quick UserVoice primer (if not already familiar): after an idea is posted, people can vote on the ideas, and the product team will be responding directly to the most popular suggestions.

We’ve imported existing feature requests from GitHub - Search for this issue there!

And don't worry, this issue will still exist on GitHub for posterity's sake.  As it’s a text-only import of the original post into UserVoice, we’ll still be keeping in mind the comments and discussion that already exist here on the GitHub issue.

GitHub will remain the channel for reporting bugs.  

Once again, this issue can now be found by searching for the title on: https://aws.uservoice.com/forums/598381-aws-command-line-interface 

-The AWS SDKs & Tools Team

**Comment by jamesls** [comment]
Based on community feedback, we have decided to return feature requests to GitHub issues.

**Comment by pierreprinetti** [comment]
Hi there. Any news on this one?

**Comment by lhanson** [comment]
AWS team: is this a feature you'd consider merging if somebody contributed the code?

**Comment by arturtamborski** [comment]
Here's a workaround until someone sends a patch:

```
export AWS_CONFIG_FILE="$XDG_CONFIG_HOME/aws/config"
export AWS_CLI_HISTORY_FILE="$XDG_DATA_HOME/aws/history"
export AWS_CREDENTIALS_FILE="$XDG_CONFIG_HOME/aws/credentials"
export AWS_WEB_IDENTITY_TOKEN_FILE="$XDG_CONFIG_HOME/aws/token"
export AWS_SHARED_CREDENTIALS_FILE="$XDG_CONFIG_HOME/aws/shared-credentials"
```

**Comment by sherabi** [comment]
It's almost end of 2019, sad to see that the XDG settings are not being recognized on a more wide basis across applications.

**Comment by lhanson** [comment]
Not to mention it's very discouraging to anyone thinking of offering a patch for anything when there are PRs sitting dormant for years adding things as simple as documentation improvements.

**Comment by lhanson** [comment]
Not to mention it's very discouraging to anyone thinking of offering a patch for anything when there are PRs sitting dormant for years providing things as simple as documentation improvements.

**Comment by timkrueger** [comment]
Should the `credentials` really stored in `$XDG_CONFIG_HOME/aws`? 

There are much users which makes usage of dotfiles repositories. Normally the `$XDG_CONFIG_HOME` is included in this repository.

Maybe the credentials should be stored in `$XDG_DATA_HOME/aws`?

**Comment by Pablo1107** [comment]
> Should the `credentials` really stored in `$XDG_CONFIG_HOME/aws`?
> 
> There are much users which makes usage of dotfiles repositories. Normally the `$XDG_CONFIG_HOME` is included in this repository.
> 
> Maybe the credentials should be stored in `$XDG_DATA_HOME/aws`?

I agree, many users include the whole $XDG_CONFIG_HOME folder in their dotfiles repo, so this would cause problems. But your recommended directory is very appropriate.

Is there any updates on this feature request?

**Comment by WhyNotHugo** [comment]
> Should the `credentials` really stored in `$XDG_CONFIG_HOME/aws`?

In all honesty, credentials should not be stored in a plain-text file in a well-known location really. You might want to consider something like [`aws-vault`](https://github.com/99designs/aws-vault).

I'm not against opposed to using `$XDG_DATA_HOME` though, just mentioning better solutions for this specific scenario.

**Comment by kdaily** [comment]
I've seen a few more requests for this change. While I do not disagree with this convention, this would be a big breaking change to many users who rely on the current configuration settings. In order to make this work, this convention would need to be adhered to by all AWS SDKs and tools. Because of this, I'm going to move this issue to the AWS SDK issue tracking repository at https://github.com/aws/aws-sdk.
